applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/differentiation_utility.cpp

```diff
@@ -0,0 +1,334 @@
+                                                Matrix& rOutput,
+                                                const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        if ( rElement.GetProperties().Has(rDesignVariable) )
+        {
+
+            // define working variables
+            Vector RHS_unperturbed;
+            Vector RHS_perturbed;
+
+            ProcessInfo copy_process_info = rCurrentProcessInfo;
+
+            // Compute RHS before perturbion
+            rElement.CalculateRightHandSide(RHS_unperturbed, copy_process_info);
+
+            if ( (rOutput.size1() != 1) || (rOutput.size2() != RHS_unperturbed.size() ) )
+                rOutput.resize(1, RHS_unperturbed.size());
+
+            // Save property pointer
+            Properties::Pointer p_global_properties = rElement.pGetProperties();
+
+            // Create new property and assign it to the element
+            Properties::Pointer p_local_property(Kratos::make_shared<Properties>(Properties(*p_global_properties)));
+            rElement.SetProperties(p_local_property);
+
+            // perturb the design variable
+            const double current_property_value = rElement.GetProperties()[rDesignVariable];
+            p_local_property->SetValue(rDesignVariable, (current_property_value + rPertubationSize));
+
+            // Compute RHS after perturbation
+            rElement.CalculateRightHandSide(RHS_perturbed, copy_process_info);
+
+            // Compute derivative of RHS w.r.t. design variable with finite differences
+            for(IndexType i = 0; i < RHS_perturbed.size(); ++i)
+                rOutput(0, i) = (RHS_perturbed[i] - RHS_unperturbed[i]) / rPertubationSize;
+
+            // Give element original properties back
+            rElement.SetProperties(p_global_properties);
+
+            //call one last time to make sure everything is as it was before TODO improve this..
+            rElement.CalculateRightHandSide(RHS_perturbed, copy_process_info);
+        }
+        else
+            if ( (rOutput.size1() != 0) || (rOutput.size2() != 0) )
+                rOutput.resize(0,0,false);
+
+        KRATOS_CATCH("");
+    }
+
+    void DifferentiationUtility::CalculateRigthHandSideDerivative(Element& rElement,
+                                                const Variable<array_1d<double,3>>& rDesignVariable,
+                                                const double& rPertubationSize,
+                                                Matrix& rOutput,
+                                                const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        #ifdef _OPENMP
+            KRATOS_ERROR_IF(omp_get_thread_num() > 0) <<
+                "DifferentiationUtility::CalculateRigthHandSideDerivative " <<
+                "is not thread safe for shape derivatives!" << omp_get_thread_num();
+        #endif
+
+        if(rDesignVariable == SHAPE)
```

Comment:
it seems to me like you will add more variables here, which make this function grow out of bounds
=> I suggest making separate functions for each variable and call them from this function (sort of a factory)

Reply:
I see your point. I the next time there will be no more possibilities. I will keep your suggestion in mind in case that the number of variables increases.